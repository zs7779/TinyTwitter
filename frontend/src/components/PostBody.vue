<template>
    <div :class="[hovering ? 'hover':'', 'p-3']" @mouseover='hovering=true' @mouseout='hovering=false'>
        <router-link :to='{name: "post", params: {username: post.author.username, postID: post.id}}' tag='div' :class="[verbose ? 'card-view':'', 'card-text']">
            <h6 class="card-title mb-1">
                <router-link :to="{ name: 'user', params: {username: post.author.username} }">{{ post.author.username }}</router-link>
                <span class="small text-muted">at {{ post.create_time }}:</span>
            </h6>
            <div v-if='post.is_comment'>
                <span class="small text-muted">Reply to</span>
                <span class="small text-muted" v-for='m in mentions' :key="m.id">
                    <router-link :to="{ name: 'user', params: {username: m.username} }">@{{ m.username }}</router-link>
                    <span> </span>
                </span>
            </div>
            <div>
                <p>{{ post.text }}</p>
            </div>
            <slot></slot>
        </router-link>
        <div v-if=buttons class="card-footer bg-transparent p-0">
            <div class="d-flex justify-content-around">
                <button type="button" v-b-modal.new-post class="btn btn-transparent px-1 py-0" title="Comment" @click="$emit('action-comment')">
                    <b-icon icon="chat" :class="[post.commented ? 'card-button-comment' : 'card-button']"></b-icon> {{ post.comment_count }}
                </button>
                <button type="button" v-b-modal.new-post class="btn btn-transparent px-1 py-0" title="Repost" @click="$emit('action-repost')">
                    <b-icon icon="arrow-repeat" :class="[post.reposted ? 'card-button-repost' : 'card-button']"></b-icon> {{ post.repost_count }}
                </button>
                <button type="button" class="btn btn-transparent px-1 py-0" title="Like" @click="$emit('action-like')">
                    <b-icon icon="heart" :class="[post.liked ? 'card-button-like' : 'card-button']"></b-icon> {{ post.like_count }}
                </button>
                <b-dropdown id="dropdown-dropup" dropup variant="btn btn-transparent px-1 py-0" no-caret title="Options">
                    <template v-slot:button-content><b-icon icon="chevron-up" class="card-button"></b-icon></template>
                    <b-dropdown-item v-if="post.owner" @click="$emit('action-delete')"><b-icon icon="x" class="card-button"></b-icon> Delete</b-dropdown-item>
                    <b-dropdown-item v-b-modal.new-post v-if="post.owner" @click="$emit('action-edit')"><b-icon icon="pencil" class="card-button"></b-icon> Edit</b-dropdown-item>
                    <b-dropdown-item><b-icon icon="share" class="card-button"></b-icon> Share</b-dropdown-item>
                </b-dropdown>
            </div>
        </div>
    </div>
</template>

<script>
export default{
    props: ['post', 'buttons', 'verbose'],
    data() {
        return {
            hovering: false,
        }
    },
    computed: {
        mentions() {
            var receipants = {};
            if (this.post.root_post) {
                receipants[this.post.root_post.author.id] = this.post.root_post.author.username;
            }
            if (this.post.parent) {
                receipants[this.post.parent.author.id] = this.post.parent.author.username;
            }
            if (this.post.mentions) {
                this.post.mentions.forEach(m => {
                    receipants[m.id] = m.username;
                })
            }
            return Object.entries(receipants).map(([k, v]) => {
                return { id: k, username: v, }
            });
        },
    }
}
</script>

<style scoped>
.hover {
    background-color: #f9f9f9;
}
.card-button {
    color: #909090;
}

.card-button-comment {
    color: darkturquoise;
}
.card-button-repost {
    color: limegreen;
}
.card-button-like {
    color: red;
}
</style>