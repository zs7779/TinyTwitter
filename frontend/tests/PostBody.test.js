import { mount } from '@vue/test-utils'
import PostBody from '../src/components/PostBody'
import mockData from './mock_data'


const factory = (post={}, buttons=false, verbose=false) => {
    return mount(PostBody, {
        stubs: ['router-link', 'b-icon', 'b-modal', 'b-dropdown', 'b-dropdown-item'],
        directives: {
            'b-modal': function() {},
        },
        propsData: {
            post: {
                ...mockData.posts[0],
                ...post,
            },
            buttons,
            verbose
        }
    });
}


describe('PostBody render', () => {
    it('Render default', () => {
        const wrapper = factory();
        expect(wrapper.html()).toMatchSnapshot();
    });
    it('Render buttons', () => {
        const wrapper = factory({}, true, false);
        expect(wrapper.html()).toMatchSnapshot();
    });
    it('Render verbose', () => {
        const wrapper = factory({}, true, false);
        expect(wrapper.html()).toMatchSnapshot();
    });
    it('Render comment', () => {
        const wrapper = factory({
            is_comment: true,
            root_post: mockData.posts[1],
            parent: mockData.comments[1],
        });
        expect(wrapper.html()).toMatchSnapshot();
    });
    it('Render repost', () => {
        const wrapper = factory({
            parent: mockData.posts[1],
        });
        expect(wrapper.html()).toMatchSnapshot();
    });
})

describe('PostBody comments', () => {
    it('Reply root', () => {
        const wrapper = factory({
            is_comment: true,
            root_post: mockData.posts[1],
        });
        expect(wrapper.find('.post-mentions').text()).toMatch(new RegExp(mockData.posts[1].author.username));
    });
    it('Reply parent', () => {
        const wrapper = factory({
            is_comment: true,
            parent: mockData.comments[1],
        }, false, false);
        expect(wrapper.find('.post-mentions').text()).toMatch(new RegExp(mockData.posts[1].author.username));
    });
    it('Not comment', () => {
        const wrapper = factory({
            parent: mockData.posts[1],
        }, false, false);
        expect(wrapper.find('.post-mentions').exists()).toBeFalsy();
    });
})