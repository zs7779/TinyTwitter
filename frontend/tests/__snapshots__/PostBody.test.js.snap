// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostBody render Render buttons 1`] = `
"<div class=\\"p-3\\">
  <router-link-stub to=\\"[object Object]\\" tag=\\"div\\" class=\\"card-text\\">
    <h6 class=\\"card-title mb-1\\">
      <router-link-stub to=\\"[object Object]\\">user0</router-link-stub> <span class=\\"small text-muted\\">at :</span>
    </h6>
    <!---->
    <div>
      <p>post0</p>
    </div>
  </router-link-stub>
  <div class=\\"card-footer bg-transparent p-0\\">
    <div class=\\"d-flex justify-content-around\\"><button type=\\"button\\" title=\\"Comment\\" class=\\"btn btn-transparent px-1 py-0\\">
        <b-icon-stub icon=\\"chat\\" class=\\"card-button\\"></b-icon-stub> 0
      </button> <button type=\\"button\\" title=\\"Repost\\" class=\\"btn btn-transparent px-1 py-0\\">
        <b-icon-stub icon=\\"arrow-repeat\\" class=\\"card-button\\"></b-icon-stub> 0
      </button> <button type=\\"button\\" title=\\"Like\\" class=\\"btn btn-transparent px-1 py-0\\">
        <b-icon-stub icon=\\"heart\\" class=\\"card-button\\"></b-icon-stub> 0
      </button>
      <b-dropdown-stub id=\\"dropdown-dropup\\" dropup=\\"\\" variant=\\"btn btn-transparent px-1 py-0\\" no-caret=\\"\\" title=\\"Options\\">
        <!---->
        <!---->
        <b-dropdown-item-stub>
          <b-icon-stub icon=\\"share\\" class=\\"card-button\\"></b-icon-stub> Share
        </b-dropdown-item-stub>
      </b-dropdown-stub>
    </div>
  </div>
</div>"
`;

exports[`PostBody render Render comment 1`] = `
"<div class=\\"p-3\\">
  <router-link-stub to=\\"[object Object]\\" tag=\\"div\\" class=\\"card-text\\">
    <h6 class=\\"card-title mb-1\\">
      <router-link-stub to=\\"[object Object]\\">user0</router-link-stub> <span class=\\"small text-muted\\">at :</span>
    </h6>
    <div class=\\"post-mentions\\"><span class=\\"small text-muted\\">Reply to</span> <span class=\\"small text-muted\\"><router-link-stub to=\\"[object Object]\\">@user0</router-link-stub> <span></span></span></div>
    <div>
      <p>post0</p>
    </div>
  </router-link-stub>
  <!---->
</div>"
`;

exports[`PostBody render Render default 1`] = `
"<div class=\\"p-3\\">
  <router-link-stub to=\\"[object Object]\\" tag=\\"div\\" class=\\"card-text\\">
    <h6 class=\\"card-title mb-1\\">
      <router-link-stub to=\\"[object Object]\\">user0</router-link-stub> <span class=\\"small text-muted\\">at :</span>
    </h6>
    <!---->
    <div>
      <p>post0</p>
    </div>
  </router-link-stub>
  <!---->
</div>"
`;

exports[`PostBody render Render repost 1`] = `
"<div class=\\"p-3\\">
  <router-link-stub to=\\"[object Object]\\" tag=\\"div\\" class=\\"card-text\\">
    <h6 class=\\"card-title mb-1\\">
      <router-link-stub to=\\"[object Object]\\">user0</router-link-stub> <span class=\\"small text-muted\\">at :</span>
    </h6>
    <!---->
    <div>
      <p>post0</p>
    </div>
  </router-link-stub>
  <!---->
</div>"
`;

exports[`PostBody render Render verbose 1`] = `
"<div class=\\"p-3\\">
  <router-link-stub to=\\"[object Object]\\" tag=\\"div\\" class=\\"card-text\\">
    <h6 class=\\"card-title mb-1\\">
      <router-link-stub to=\\"[object Object]\\">user0</router-link-stub> <span class=\\"small text-muted\\">at :</span>
    </h6>
    <!---->
    <div>
      <p>post0</p>
    </div>
  </router-link-stub>
  <div class=\\"card-footer bg-transparent p-0\\">
    <div class=\\"d-flex justify-content-around\\"><button type=\\"button\\" title=\\"Comment\\" class=\\"btn btn-transparent px-1 py-0\\">
        <b-icon-stub icon=\\"chat\\" class=\\"card-button\\"></b-icon-stub> 0
      </button> <button type=\\"button\\" title=\\"Repost\\" class=\\"btn btn-transparent px-1 py-0\\">
        <b-icon-stub icon=\\"arrow-repeat\\" class=\\"card-button\\"></b-icon-stub> 0
      </button> <button type=\\"button\\" title=\\"Like\\" class=\\"btn btn-transparent px-1 py-0\\">
        <b-icon-stub icon=\\"heart\\" class=\\"card-button\\"></b-icon-stub> 0
      </button>
      <b-dropdown-stub id=\\"dropdown-dropup\\" dropup=\\"\\" variant=\\"btn btn-transparent px-1 py-0\\" no-caret=\\"\\" title=\\"Options\\">
        <!---->
        <!---->
        <b-dropdown-item-stub>
          <b-icon-stub icon=\\"share\\" class=\\"card-button\\"></b-icon-stub> Share
        </b-dropdown-item-stub>
      </b-dropdown-stub>
    </div>
  </div>
</div>"
`;
